### ISSUE TITLE
Unexpected gradient of sums does not match sum of gradients

### ISSUE THREAD

**Issue by rubenvereecken** [pytorch/pytorch#26161]
## 🐛 Bug

To preface, this started as [a post on the forum](https://discuss.pytorch.org/t/unexpected-gradient-of-sum-doesnt-match-sum-of-gradients/55790). However, I now think it's a bug or unexpected behaviour and merits a discussion here.

My problem is straightforward: I have a gradient of a sum that turns out 0 while I think it shouldn’t be, and it doesn’t match the sum of the gradients separately.

## To Reproduce
I have a simple example to illustrate. We’re in the probability domain, or rather, the logprob domain, so this is the straightforward approximation of probabilities using a softmax on top of a linear layer. Or rather, a logsoftmax.

```
import torch
import torch.nn as nn
import torch.nn.functional as F

encoder = nn.Linear(1, 2, bias=False)
nn.init.zeros_(encoder.weight.data)
# Fabricate some one-hot encoded data of batch size 1, good enough
input = torch.ones((1,1))
```
As I said, I have the sum of gradients, or rather a difference really. Below I’ll show the gradients of the linear layer’s weights after backpropagating first the difference, then the summands themselves separately.
```
logits = encoder.forward(input)
norm = torch.logsumexp(logits, dim=1, keepdim=True)
# --- This is the sum producing strange gradients!
(logits-norm).sum().backward()
print(encoder.weight.grad)
nn.init.zeros_(encoder.weight.grad.data)

logits = encoder.forward(input)
logits.sum().backward()
print(encoder.weight.grad)
nn.init.zeros_(encoder.weight.grad.data)

logits = encoder.forward(input)
norm = torch.logsumexp(logits, dim=1, keepdim=True)
norm.sum().backward()
print(encoder.weight.grad)
```
This prints, in order, annotated
```
# grad of logits - norm
tensor([[0.],
        [0.]])
# grad of logits
tensor([[1.],
        [1.]])
# grad of norm
tensor([[0.5000],
        [0.5000]])
```
## Expected behavior
The gradients for logits and norm are exactly what I expected. I expected the gradient for logprobs to be the difference of the gradients of logits and norm, ie `[[.5, .5]]`, but that’s not the case. Instead it’s zeros.

## What I think happened
Since there are two outputs, we can look at the derivative of the linear layer for each output separately.
```
logprobs[:,0].sum().backward()
encoder.weight.grad ==  tensor([[ 0.5000], [-0.5000]])
logprobs[:,1].sum().backward()
encoder.weight.grad ==  tensor([[ -0.5000], [0.5000]])
```
Summed up, these gradients would all become 0. While it can be understood this way, it doesn't seem like the right behaviour. I also don't know how to work around this consistently. If it does work as intended, how should I correctly go about creating the behaviour I expected?

## Environment

Versions of relevant libraries:
```
[pip] numpy==1.16.3
[pip] torch==1.2.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.3                      199  
[conda] mkl_fft                   1.0.12           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] torch                     1.2.0                    pypi_0    pypi
```

**Comment by rubenvereecken** [comment]
Alright I believe I better understand the big picture.

The sum of the gradients doesn't match, because I didn't actually look at all the gradients. Instead I looked at the gradients summed over outputs. If I had instead looked at the Jacobian of the summands along the way, I would see that the sum of those Jacobians would match the final Jacobian. This final Jacobian, when summed along the output axis, will be a 0 vector. 

To summarise, my confusion resulted from the fact that I didn't simply look at the gradient for a single output, but for multiple outputs. PyTorch simply adds up those gradients, which resulted in my case in a vector of zeros.
